### DISCUSSION TITLE
How to globally capture matched params before / after route change?

### DISCUSSION THREAD

**Discussion by borie88** [TanStack/router#128] [Q&A] [answered]
Is there a way to listen for route changes and do something with the matched information like search / params? Vue router has a `beforeEach` callback on a router instance that makes it really simple to get information about the currently resolved route.

In my case, I'm trying to implement a flow that subscribes to a websocket based on the ':id' of something in the URL, but there are multiple routes / components that use this ':id', so it makes sense to put this logic at the top-level of the router.

I thought the Root component used as a child of the Router component in some of the examples might be able to do that, but all hooks return information resolved from root '/', which makes sense.

**Comment by tannerlinsley** [comment, marked as answer]
You could try using a wildcard with useMatchRoute from the root. Or you could just listen to useRouter().state.matches, and use the params from the last match (it would have all of them)

**Comment by tannerlinsley** [comment]
When I say listen, I mean useEffect and friends

**Comment by borie88** [comment]
useEffect with useRouter().state.matches works like a charm. thanks!
